### DISCUSSION TITLE
Sending h264 frames

### DISCUSSION THREAD

**Discussion by HigerSkill** [paullouisageneau/libdatachannel#1286] [Q&A] [answered]
I'm new to WebRTC and video encoding. I'm trying to encode images in `cv::Mat` to h264 and send them via WebRTC to the frontend, similar to the media-sender example. I'm facing a problem where, although I encode the image and set up the h264 packetizer, when I send the video, there are no errors, but no video appears on the frontend.

```C++
        rtc::Description::Video media("video", rtc::Description::Direction::SendOnly);
        media.addH264Codec(102);
        constexpr rtc::SSRC ssrc = 42;
        media.addSSRC(ssrc, "video");
        video_track = peer->addTrack(media);
        auto rtpConfig = std::make_shared<rtc::RtpPacketizationConfig>(ssrc, "video", 102,
                                                                          rtc::H264RtpPacketizer::defaultClockRate);
        auto packetizer = std::make_shared<rtc::H264RtpPacketizer>(rtc::NalUnit::Separator::LongStartSequence, rtpConfig);
        auto srReporter = std::make_shared<rtc::RtcpSrReporter>(rtpConfig);
        packetizer->addToChain(srReporter);
        auto nackResponder = std::make_shared<rtc::RtcpNackResponder>();
        packetizer->addToChain(nackResponder);
        video_track->setMediaHandler(packetizer);
 ```
Why might my RTP packets be ignored?

I found related problem in discussions https://github.com/paullouisageneau/libdatachannel/discussions/699
I have the same problem, but there no answers

**Comment by vvalentinich** [comment]
Do you encode to h264 properly (you can write raw h264 frames to a file for testing)? Do you observe suspicious records in libdatachannel's log? If a browser is used on receivers side, do you see something wrong from webrtc's details (Chrome - chrome://webrtc-internals; Edge    - edge://webrtc-internals; Firefox - about:webrtc)?

**Reply by HigerSkill** [reply]
The frames are saved to the file successfully and can be viewed, everything is correct

```C++
        cv::Mat frame(msg->height, msg->width, CV_8UC3, msg->data.data());
        cv::Mat yuv_frame;
        cvtColor(frame, yuv_frame, cv::COLOR_RGB2YUV_I420);

        x264_picture_t pic_in, pic_out;
        x264_picture_alloc(&pic_in, X264_CSP_I420, WIDTH, HEIGHT);
        memcpy(pic_in.img.plane[0], yuv_frame.data, WIDTH * HEIGHT);
        memcpy(pic_in.img.plane[1], yuv_frame.data + WIDTH * HEIGHT, WIDTH * HEIGHT / 4);
        memcpy(pic_in.img.plane[2], yuv_frame.data + WIDTH * HEIGHT * 5 / 4, WIDTH * HEIGHT / 4);

        x264_nal_t *nals;
        int i_nals;
        int frame_size = x264_encoder_encode(encoder_, &nals, &i_nals, &pic_in, &pic_out);
```
There are no errors, I didn't find anything suspicious on the receivers  side either.
I launched the streamer example and replaced the frames with my own, but they are not displayed either, although there are no errors. It seems that the problem is really in the encoding in h264, but it is not clear what exactly the problem is, if the frames are correct when saved to a file. 

Perhaps you know how to properly encode frames without using cli from gstreamer or ffmpeg, now I use x264 library.

**Reply by vvalentinich** [reply]
I used encoding to H264 on Windows only (IMFTransform,...). Do you see on receivers side that data actually arrived? If yes, a browser will show a statistic related to webrtc (bitrate,...). Also I would check timestamping of video frames.

**Reply by HigerSkill** [reply]
![image](https://github.com/user-attachments/assets/fc5226d5-b085-4b51-ba05-a6c2afa291cb)
![image](https://github.com/user-attachments/assets/d9850077-42cd-4459-8dd7-f1338973ea01)
![image](https://github.com/user-attachments/assets/7313dedd-c3b8-48d1-996c-81bfcced5354)


I see that traffic is passing and packets are being sent, but I don't see where to look for bitrate and video parameters

Which timestamps should I check? Do I need to do something with the timestamp when encoding?

**Reply by vvalentinich** [reply]
I meant things like rtpConfig->timestamp (from streamer example). As for pictures, I would look carefully at "Stats graphs for inbound-rtp (kind=video...".

**Reply by HigerSkill** [reply]
![image](https://github.com/user-attachments/assets/c5660cc9-1271-4121-b5e8-e63cff7addac)

I didn't change anything in the example, I just output the `rtpConfig->timestamp`, for some reason it's incorrect.
```C++
Timestamp: 3437318314
Timestamp: 3321433375
Timestamp: 3437319274
Timestamp: 3437320234
Timestamp: 3321436375
Timestamp: 3437321194
```

I manually set `rtpConfig->startTimestamp` to the current time, it didn't help either.

For the original frames the timestamps are also strange, but they work:
```C++
Timestamp: 4143417708
Timestamp: 4143418668
Timestamp: 2680257504
Timestamp: 4143419628
Timestamp: 4143420588
Timestamp: 2680260504
Timestamp: 4143421548
Timestamp: 2680263504
Timestamp: 4143422508
Timestamp: 4143423468
```

**Reply by paullouisageneau** [reply, marked as answer]
The timestamps look fine: the difference between two audio timestamps is 960 (20ms at 48 KHz), and the difference between video timestamps is 3000 (around 31.25ms at 96KHz).

As you can see on the Chrome graph, it receives packets but no valid frames. This looks like an encoding issue. How do you bundle NAL units together into frames? It has to be consistent with the packetizer separator setting.

**Reply by HigerSkill** [reply]
@paullouisageneau @vvalentinich Thanks for the help, the problem was indeed in `NalUnit::Separator` and it was necessary to correctly assemble the frame
